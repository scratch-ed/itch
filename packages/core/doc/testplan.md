# Test plans

Test plans are used to test submissions for Scratch exercises.
While intimate knowledge of the judge is not required, it can be useful to look at the source code in case of doubt.

## Structure

A test plan is split into three main stages:

1. `beforeExecution` - Executed before the Scratch VM is run.
2. `duringExecution` - Used to drive the test plan.
3. `afterExecution` - Executed after the solution has been executed.

In the `beforeExecution` stage, you have access to the solution and the starter version of the exercise.
While the Scratch VM has not yet run, you do have access to all sprites and code.
The access to the starter version of the exercise is useful to perform checks to verify that predefined blocks did not change.

The stages are implemented as a function, which will be called by the judge at the right moment.
For example, the `beforeExecution` is as follows:

```javascript
/**
 * @param {Evaluation} e
 */
function beforeExecution(e) {
  // Access the submission
  e.log.submission;
  // Access the starter version.
  e.log.template;
}
```

The signature of the other stages is the same.

The `duringExecution` stage is mainly used to drive the Scratch VM by using the scheduler.
The scheduler takes the place of the human;
it will click on sprites, it will send broadcasts, do key presses, etc.

For example, it is very common to send the green flag broadcast.

Finally, the `afterExecution` stage is run after the submission has been run.
Here, you have access to the captured log, so tests on what happens are possible.

## Before execution

This phase is meant for static tests, where execution of the submission is not necessary.

One of the most common tests is the test for the predefined blocks.
Students often change the predefined blocks or change the code of other sprites.
Since other tests rely on the predefined blocks to work, we need to make sure they actually work.

Since this is used so often, there is built-in support for checking the predefined blocks.
The documentation is available at TODO.
An example of this is the following snippet:

```javascript
function beforeExecution(e) {
  Itch.checkPredefinedBlocks(
    {
      hats: {
        Planet: [stack(whenIReceive('Start'), setSizeTo(anything()))]
      },
    },
    e
  );
}
```

The built-in checks work by comparing the blocks of the submission to the blocks of the template.
However, since students must add blocks somewhere, we must define where students are allowed to add blocks.

In the example above, we allow blocks to be added to the "Planet" sprite,
in a block stack starting with the hat _When I receive broadcast "Start"_,
followed by a block "Set size to", with any argument.


## During execution

Here, the main purpose is to simulate user interaction.
For example, you often want to press the green flag.

A simple example is:

```javascript
function duringExecution(e) {
  e.scheduler
    .greenFlag(false)
    .sendBroadcast('Start', false)
    .wait(1000);
}
```

In this example, we press the green flag (and don't wait for execution to end),
then send a broadcast "Start" (again not waiting for the execution to end),
and finally we wait 1000 ms (so 1 second).

The reference documentation for the scheduler is available at [].


## After execution

After the VM has stopped executing, you can access the log to inspect what happened (e.g. postmortem tests).

See the docs of the log for more details on available data.

## Globals

The test plan is a traditional javascript file, not a module (unfortunately).
This means you cannot import functions or other constructs.
Therefore, some global methods are available.
These are documented in the type declaration at [].

## Matching blocks

In various places of the test plan, you might want to match blocks.
To make things easier, a block API exists;
a kind of DSL to match with blocks.

Most matching functions accept a pattern.
A pattern supports the usual techniques:

- Wildcards (with `anything()`).
- Choices (with a list).
- Exact matches.

For example, consider the following block stack:

![image](./stack.png)

Can be represented as such by the API:

```javascript
stack(
  whenIReceive('Start'),
  setEffectTo(transparent(), 0),
  repeat(15, stack(
    changeSizeBy(3),
    changeYBy(-2), 
  )),
  repeat(20, stack(
    changeEffectBy(transparent(), 5),
    changeSizeBy(3)
  )),
  hide()
)
```


For most blocks, an equivalent function exists (see the source file).
Three special functions are:

- `stack()`: Indicate a block stack.
- `anything()`: Matches anything.
- `nothing()`: Matches nothing.
