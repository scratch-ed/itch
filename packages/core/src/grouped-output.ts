/**
 * @fileOverview A way to group tests in a flexible hierarchy. In the long term,
 *   this will replace the existing groups.
 */

/**
 * Handle outputting. By default, all output is sent to stderr.
 * You can overwrite this by setting a global `handleOut` callback,
 * which will be used instead if present.
 */
function toOutput(output: Record<string, unknown>): void {
  // @ts-ignore
  if (typeof window.handleOut !== 'undefined') {
    // @ts-ignore
    window.handleOut(output);
  } else {
    console.log(output);
  }
}

export enum Status {
  TimeLimit = 'time limit exceeded',
  Runtime = 'runtime error',
  Wrong = 'wrong',
  Correct = 'correct',
}

export enum Format {
  Text = 'text',
  Html = 'html',
}

export interface OutputHandler {
  (obj: Record<string, unknown>): void;
}

export interface Message {
  format: Format;
  description: string;
}

export enum Visibility {
  Show = 'show',
  Collapse = 'collapse',
}

/**
 * Manages the output for the Dodona-inspired format.
 *
 * While this class is exposed in testplans, in most cases
 * you should use the high-level testplan API instead of this one.
 *
 * ### Dodona format
 *
 * Some more information on the Dodona format. The format is a partial format.
 * The judge basically sends updates to the test result state via commands, e.g.
 * "start testcase X", "start test Y", "close testcase X", etc.
 *
 * For ease of use, the result manager will automatically open higher levels when
 * opening lower levels. For example, if you open a testcase without opening a
 * context first, the result manager will do so for you. Previous levels are also
 * closed when appropriate. For example, when starting a new tab, all previous tabs
 * will be closed.
 *
 * There is one exception: a test. If an open test is detected, an error will be thrown,
 * as the result manager has no way of knowing if the test is successful or not.
 */
export class GroupedResultManager {
  private readonly out: OutputHandler;
  private hasOpenJudgement = false;
  private openGroups = 0;
  private hasOpenTest = false;
  private isFinished = false;
  private escalation?: Status;

  constructor(out: OutputHandler = toOutput) {
    this.out = out;
  }

  /**
   * Start the judgement.
   */
  startJudgement(): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to start judgement after judgement has been completed. Ignoring.',
      );
      return;
    }
    this.out({ command: 'start-judgement', version: 2 });
    this.hasOpenJudgement = true;
  }

  /**
   * Close the judgement. This will finish the judge, meaning all future
   * output is ignored.
   */
  closeJudgement(): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to close judgement after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to close judgement while none is open. Ignoring.');
      return;
    }
    while (this.openGroups > 0) {
      console.warn('Unclosed groups, force closing them...');
      this.closeGroup();
    }
    this.out({ command: 'close-judgement' });
    this.hasOpenJudgement = false;
    this.isFinished = true;
  }

  /**
   * Start a new test group.
   *
   * Each group needs a name, which is user visible. You can also optionally
   * provide the name of a sprite, which allows for a better display, for
   * example by showing the sprite or linking to it. You must use the sprite's
   * name.
   *
   * @param name Name of the test group.
   * @param [sprite] Name of optionally linked sprite.
   */
  startGroup(name: string, sprite?: string): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to open group after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (!this.hasOpenJudgement) {
      this.startJudgement();
    }
    this.openGroups++;
    this.out({ command: 'start-group', name: name, sprite: sprite });
  }

  /**
   * Close a group.
   *
   * @param summary
   * @param visibility
   */
  closeGroup(summary?: string | Message, visibility: Visibility = Visibility.Show): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to close group after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (this.openGroups <= 0) {
      console.warn('Attempting to close tab while none is open. Ignoring.');
      return;
    }

    this.openGroups--;
    this.out({ command: 'close-group', summary: summary, visibility: visibility });
  }

  startTest(name = `Test ${Math.random()}`): void {
    if (this.isFinished) {
      console.warn('Attempting to start test after judgement has been completed.');
      return;
    }
    if (this.hasOpenTest) {
      console.warn(
        'Attempting to start new test while open test exists, closing it as wrong.',
      );
      this.closeTest(Status.Wrong);
    }
    if (this.openGroups <= 0) {
      console.warn('Attempting to start test while no group is open.');
      this.startGroup(`Groep ${Math.random()}`);
    }
    this.out({
      command: 'start-test',
      name: name,
    });
    this.hasOpenTest = true;
  }

  closeTest(status: Status, description?: string): void {
    if (this.isFinished) {
      console.warn('Attempting to close test after judgement has been completed.');
      return;
    }
    if (!this.hasOpenTest) {
      console.warn('Attempting to close test while none is open. Ignoring.');
      return;
    }
    this.out({
      command: 'close-test',
      status: status,
      description: description,
    });
    this.hasOpenTest = false;
  }

  appendMessage(message: string): void {
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to append message while no judgement is open. Ignoring.');
      return;
    }
    this.out({ command: 'append-message', message: message });
  }

  appendDiff(expected: string, actual: string | null = null): void {
    if (!this.hasOpenTest) {
      console.warn('Attempting to append diff while no test is open. Ignoring.');
      return;
    }
    this.out({ command: 'append-diff', expected: expected, actual: actual });
  }

  escalateStatus(status: Status | string): void {
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to escalate status of closed judgement. Ignoring.');
      return;
    }
    this.escalation = status as Status;
    this.out({ command: 'escalate-status', status: status });
  }

  whyEscalate(): Status | undefined {
    return this.escalation;
  }
}
