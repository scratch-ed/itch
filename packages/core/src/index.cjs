// Wrapper for common js modules.
const result = require('./evaluation.js');

module.exports = result.run;
