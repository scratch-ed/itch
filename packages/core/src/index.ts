// Wrapper for common js modules.
import { run } from './evaluation';
export default run;
