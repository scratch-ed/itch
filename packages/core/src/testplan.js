/**
 * @file This file contains the testplan API, i.e. most of the stuff
 * you use when writing a test plan. This API is inspired by Jest, so
 * if you are familiar, it should be fairly easy to pick up.
 *
 * ## Structure
 *
 * Itch provides 4 levels of groupings for tests:
 *
 * 1. `tab`
 * 2. `describe`
 * 3. `test`
 * 4. `expect`
 *
 * When starting from the bottom, we begin simple:
 *
 * 1. The `expect` is used to compare two values. It does not have a name itself,
 *    but you can provide a custom error message. This is only shown when the
 *    assertion fails. (The values are always passed as well).
 * 2. The `test` is the lowest level with a name. It groups a bunch of related
 *    `expect` statements.
 * 3. The `describe` directive groups a bunch of related tests, e.g. for one sprite.
 * 4. The `tab` groups a bunch of `describe` statements. These are mainly for UI purposes.
 */
import isEqual from 'lodash/isEqual';
import { CORRECT, WRONG } from './output.js';
import { castCallback } from './utils.js';

export class FatalErrorException extends Error {}

class GenericMatcher {
  constructor(context, actual) {
    /** @type {Context} */
    this.context = context;
    this.actual = actual;
    /** @type {function(any, any):string} */
    this.errorMessage = null;
    /** @type {function(any, any):string} */
    this.successMessage = null;
    /** @type {boolean} */
    this.terminate = false;
  }

  /**
   * Post the result.
   *
   * @param {boolean} accepted - If the property satisfies the condition.
   * @param {string} [errorMessage] - Default error message.
   * @param {string} [successMessage] - Optional success message.
   *
   * @private
   */
  out(accepted, errorMessage = undefined, successMessage = undefined) {
    this.context.output.startTest(this.expected);
    const status = accepted ? CORRECT : WRONG;

    if (accepted) {
      const message = this.successMessage
        ? this.successMessage(this.expected, this.actual)
        : successMessage;
      if (message) {
        this.context.output.appendMessage(message);
      }
    } else {
      const message = this.errorMessage
        ? this.errorMessage(this.expected, this.actual)
        : errorMessage;
      if (message) {
        this.context.output.appendMessage(message);
      }
    }

    this.context.output.closeTest(this.actual, accepted, status);

    if (!accepted && this.terminate) {
      throw new FatalErrorException();
    }
  }

  /**
   * Allows setting an error message.
   *
   * @param {string|function(any,any):string} message
   *
   * @return {GenericMatcher}
   * @deprecated
   */
  withError(message) {
    this.errorMessage = castCallback(message);
    return this;
  }

  /**
   * Allows setting an error message.
   *
   * @param {string|function(any,any):string} message
   *
   * @return {GenericMatcher}
   * @deprecated
   */
  withSuccess(message) {
    this.successMessage = castCallback(message);
    return this;
  }

  /**
   * Combines withSuccess & withError
   *
   * @param {{[correct]:string|function():string, [wrong]:string|function():string}} messages
   *
   * @return {GenericMatcher}
   */
  with(messages) {
    this.successMessage = castCallback(messages.correct);
    this.errorMessage = castCallback(messages.wrong);
    return this;
  }

  /**
   * Mark this test as fatal: if it fails, the testplan will stop.
   *
   * @return {GenericMatcher}
   */
  fatal() {
    this.terminate = true;
    return this;
  }

  /**
   * Compares two values for equality.
   *
   * Most types of objects should be supported:
   *
   * - If both values are numbers, `numericEquals` is used, which supports floats.
   * - Otherwise, the `isEqual` function from lodash is used. Quoting their docs:
   *
   *   > This method supports comparing arrays, array buffers, booleans,
   *   > date objects, error objects, maps, numbers, `Object` objects, regexes,
   *   > sets, strings, symbols, and typed arrays. `Object` objects are compared
   *   > by their own, not inherited, enumerable properties. Functions and DOM
   *   > nodes are **not** supported.
   *
   *
   * @param expected
   */
  toBe(expected) {
    this.expected = expected;
    if (typeof this.actual === 'number' && typeof expected === 'number') {
      this.out(
        numericEquals(this.actual, expected),
        `Expected ${JSON.stringify(expected)} but got ${JSON.stringify(
          this.actual,
        )}`,
      );
    } else {
      this.out(
        isEqual(this.actual, expected),
        `Expected ${JSON.stringify(expected)} but got ${JSON.stringify(
          this.actual,
        )}`,
      );
    }
  }

  /**
   * Compares two values for equality.
   *
   * Most types of objects should be supported:
   *
   * - If both values are numbers, `numericEquals` is used, which supports floats.
   * - Otherwise, the `isEqual` function from lodash is used. Quoting their docs:
   *
   *   > This method supports comparing arrays, array buffers, booleans,
   *   > date objects, error objects, maps, numbers, `Object` objects, regexes,
   *   > sets, strings, symbols, and typed arrays. `Object` objects are compared
   *   > by their own, not inherited, enumerable properties. Functions and DOM
   *   > nodes are **not** supported.
   *
   *
   * @param expected
   */
  toNotBe(expected) {
    this.expected = expected;
    if (typeof this.actual === 'number' && typeof expected === 'number') {
      this.out(
        !numericEquals(this.actual, expected),
        `Expected ${expected?.toString()} but got ${this.actual?.toString()}`,
      );
    } else {
      this.out(
        !isEqual(this.actual, expected),
        `Expected ${expected?.toString()} but got ${this.actual?.toString()}`,
      );
    }
  }
}

class ExpectLevel {
  /**
   * @param {Context} context
   * @package
   */
  constructor(context) {
    this.context = context;
  }

  /**
   * Start an assertion be providing a value.
   *
   * You can optionally provide a custom error message.
   *
   * @param {*} value
   *
   * @return {GenericMatcher}
   */
  expect(value) {
    return new GenericMatcher(this.context, value);
  }

  /**
   * Add a test that will always be accepted.
   */
  accept() {
    this.context.output.startTest(true);
    this.context.output.closeTest(true, true);
  }
}

class TestLevel {
  /**
   * @param {Context} context
   * @package
   */
  constructor(context) {
    this.context = context;
  }

  /**
   * Check some properties as part of the same test.
   *
   * ### Test vs expect
   *
   * A good guideline to decide whether testing multiple properties
   * should happen in the same file or not: if it needs a name, it should
   * be a separate test.
   *
   * For example, if you are testing that a sprite moves down when pressing a key,
   * you might have one test, with two properties: one for x and one for y.
   *
   * On the other hand, you might want to have multiple tests: one for the position,
   * one for the orientation, etc.
   *
   * This level results in a `testcase` in the output format.
   *
   * @param {string} name
   * @param {function(out:ExpectLevel)} block
   */
  test(name, block) {
    this.context.output.startTestcase(name);
    block(new ExpectLevel(this.context));
    this.context.output.closeTestcase();
  }
}

class DescribeLevel extends TestLevel {
  /**
   * Groups a bunch of related tests.
   *
   * This level results in a `context` in the output format.
   *
   * @param {string} name - Either the name or the function.
   * @param {function(TestLevel)} block - The function if a name is passed.
   */
  describe(name = undefined, block = undefined) {
    this.context.output.startContext(name);
    block(this);
    this.context.output.closeContext();
  }
}

export class TabLevel extends DescribeLevel {
  /**
   * Run the tests in the block inside the tab.
   *
   * This level results in a `tab` in the output format.
   *
   * @param {string} name
   * @param {function(DescribeLevel)} block
   */
  tab(name, block) {
    this.context.output.startTab(name);
    block(this);
    this.context.output.closeTab();
  }
}
