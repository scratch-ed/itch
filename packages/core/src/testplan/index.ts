/**
 * @file This file contains the testplan API, i.e. most of the stuff
 * you use when writing a test plan. This API is inspired by Jest, so
 * if you are familiar, it should be fairly easy to pick up.
 *
 * ## Structure
 *
 * Itch provides 4 levels of groupings for tests:
 *
 * 1. `tab`
 * 2. `describe`
 * 3. `test`
 * 4. `expect`
 *
 * When starting from the bottom, we begin simple:
 *
 * 1. The `expect` is used to compare two values. It does not have a name itself,
 *    but you can provide a custom error message. This is only shown when the
 *    assertion fails. (The values are always passed as well).
 * 2. The `test` is the lowest level with a name. It groups a bunch of related
 *    `expect` statements.
 * 3. The `describe` directive groups a bunch of related tests, e.g. for one sprite.
 * 4. The `tab` groups a bunch of `describe` statements. These are mainly for UI purposes.
 */
import type VirtualMachine from 'itch-scratch-vm-types';
import type BlockUtility from 'itch-scratch-vm-types/types/engine/block-utility';

export class FatalErrorException extends Error {}

/**
 * EXPERIMENTAL!
 *
 * Intercepts wait blocks in procedure definitions in the given sprite for the
 * given amount and ignores them.
 *
 * @param {VirtualMachine} vm
 * @param {string} sprite
 */
export function ignoreWaitInProcedureFor(vm: VirtualMachine, sprite: string): void {
  const original = vm.runtime._primitives.control_wait;
  vm.runtime._primitives.control_wait = (args: unknown, util: BlockUtility) => {
    // Big hack to ignore wait in movement steps.
    if (util.thread!.target?.getName() === sprite) {
      const glowId = util.thread!.blockGlowInFrame;
      if (glowId) {
        let current = util.thread!.blockContainer.getBlock(glowId);
        while (current?.parent !== null) {
          current = util.thread!.blockContainer.getBlock(current.parent);
        }
        if (current?.opcode === 'procedures_definition') {
          console.log(`Skipping ... ${current?.opcode}`);
          return;
        }
      }
    }

    original(args, util);
  };
}

interface Range {
  min: number;
  max: number;
}

export function asRange(range: number | Range): Range {
  if (typeof range === 'number') {
    return { min: range, max: range };
  } else {
    return range;
  }
}
