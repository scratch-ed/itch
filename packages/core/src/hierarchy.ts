/**
 * @file This file contains the testplan API, i.e. most of the stuff
 * you use when writing a test plan. This API is inspired by Jest, so
 * if you are familiar, it should be fairly easy to pick up.
 *
 * ## Structure
 *
 * Itch provides 4 levels of groupings for tests:
 *
 * 1. `group`   -> a group in the output
 * 2. `test` -> a test in the output
 *
 * When starting from the bottom, we begin simple:
 *
 * 1. The `expect` is used to compare two values. It does not have a name itself,
 *    but you can provide a custom error message. This is only shown when the
 *    assertion fails. (The values are always passed as well).
 * 2. The `test` is the lowest level with a name. It groups a bunch of related
 *    `expect` statements.
 * 3. The `describe` directive groups a bunch of related tests, e.g. for one sprite.
 * 4. The `tab` groups a bunch of `describe` statements. These are mainly for UI purposes.
 */
import isEqual from 'lodash-es/isEqual';
import { castCallback, numericEquals, stringify } from './utils';
import { GroupedResultManager, Status, Visibility } from './grouped-output';

export class FatalErrorException extends Error {}

export class RuntimeException extends Error {}

interface OutputCallback {
  (accepted: boolean, expected?: unknown, actual?: unknown): void;
}

class Matcher {
  private readonly callback: OutputCallback;
  private readonly actual: unknown;

  constructor(callback: OutputCallback, actual: unknown) {
    this.callback = callback;
    this.actual = actual;
  }

  private accepted(expected: unknown): boolean {
    if (typeof this.actual === 'number' && typeof expected === 'number') {
      return numericEquals(this.actual, expected);
    } else {
      return isEqual(this.actual, expected);
    }
  }

  /**
   * Compares two values for equality.
   *
   * Most types of objects should be supported:
   *
   * - If both values are numbers, `numericEquals` is used, which supports floats.
   * - Otherwise, the `isEqual` function from lodash is used. Quoting their docs:
   *
   *   > This method supports comparing arrays, array buffers, booleans,
   *   > date objects, error objects, maps, numbers, `Object` objects, regexes,
   *   > sets, strings, symbols, and typed arrays. `Object` objects are compared
   *   > by their own, not inherited, enumerable properties. Functions and DOM
   *   > nodes are **not** supported.
   */
  toBe(expected: unknown): void {
    this.callback(this.accepted(expected), expected, this.actual);
  }

  /**
   * Compares two values for equality.
   *
   * Most types of objects should be supported:
   *
   * - If both values are numbers, `numericEquals` is used, which supports floats.
   * - Otherwise, the `isEqual` function from lodash is used. Quoting their docs:
   *
   *   > This method supports comparing arrays, array buffers, booleans,
   *   > date objects, error objects, maps, numbers, `Object` objects, regexes,
   *   > sets, strings, symbols, and typed arrays. `Object` objects are compared
   *   > by their own, not inherited, enumerable properties. Functions and DOM
   *   > nodes are **not** supported.
   */
  toNotBe(expected: unknown): void {
    this.callback(!this.accepted(expected), expected, this.actual);
  }
}

export type MessageCallback = (..._args: unknown[]) => string;

interface CorrectMessage {
  correct: string | MessageCallback;
  wrong?: string | MessageCallback;
}

interface WrongMessage {
  correct?: string | (() => string);
  wrong: string | (() => string);
}

export type Messages = CorrectMessage | WrongMessage;

class TestOptions {
  private errorMessage?: MessageCallback;
  private correctMessage?: MessageCallback;
  private terminate = false;
  private enableDiff = false;

  constructor(private readonly resultManager: GroupedResultManager) {}

  /**
   * Make this test fatal: if it fails, execution is stopped.
   */
  fatal(): TestOptions {
    this.terminate = true;
    return this;
  }

  /**
   * Enable the diff output.
   */
  diff(): TestOptions {
    this.enableDiff = true;
    return this;
  }

  /**
   * Execute the test.
   *
   * @param accepted - If the test is correct or not.
   */
  do(accepted: boolean): void {
    if (this.enableDiff) {
      throw new RuntimeException('Cannot request diff if no values are given.');
    }

    this.out(accepted);
  }

  /**
   * Execute the test with
   * @param actual
   */
  expect(actual: unknown): Matcher {
    // We pass a method, so set the "this" correctly.
    return new Matcher(this.out.bind(this), actual);
  }

  feedback(messages: Messages): TestOptions {
    this.correctMessage = castCallback(messages.correct);
    this.errorMessage = castCallback(messages.wrong);
    return this;
  }

  private out(accepted: boolean, expected?: unknown, actual?: unknown) {
    const status = accepted ? Status.Correct : Status.Wrong;
    const messageCallback = accepted ? this.correctMessage : this.errorMessage;
    const summary = messageCallback?.(expected, actual);

    if (this.enableDiff) {
      this.resultManager.appendDiff(stringify(expected), stringify(actual));
    }

    this.resultManager.closeTest(status, summary);

    if (!accepted && this.terminate) {
      throw new FatalErrorException();
    }
  }
}

export interface GroupOptions {
  sprite?: string;
  summary?: string;
  visibility?: Visibility | 'show' | 'collapse';
}

export interface SpriteGroupOptions extends GroupOptions {
  sprite: string;
}

export class GroupLevel {
  constructor(protected readonly resultManager: GroupedResultManager) {}

  /**
   * Start a test.
   * @param name - Optional name of the test.
   */
  test(name?: string): TestOptions {
    if (name) {
      this.resultManager.startTest(name);
    } else {
      this.resultManager.startTest();
    }

    return new TestOptions(this.resultManager);
  }

  /**
   * Groups a bunch of related tests.
   *
   * This level results in a `context` in the output format.
   *
   * @param name - Either the name or the function.
   * @param options - The name of the sprite that is linked to this group.
   * @param block - The function if a name is passed.
   */
  group(name: string, options: GroupOptions | string, block: () => void): void;

  /**
   * Groups a bunch of related tests.
   *
   * This level results in a `context` in the output format.
   *
   * @param name - Either the name or the function.
   * @param block - The function if a name is passed.
   */
  group(name: string, block: () => void): void;

  /**
   * Start a group for a certain sprite.
   *
   * @param options - At least the sprite should be given.
   * @param block - The block.
   */
  group(options: SpriteGroupOptions, block: () => void): void;

  group(
    name: string | SpriteGroupOptions,
    spriteOrBlock: string | GroupOptions | (() => void),
    block?: () => void,
  ): void {
    if (typeof spriteOrBlock === 'string') {
      spriteOrBlock = { sprite: spriteOrBlock };
    }
    if (typeof name !== 'string') {
      spriteOrBlock = name;
      name = `Testen voor ${name.sprite}`;
    }
    if (typeof spriteOrBlock !== 'object') {
      block = spriteOrBlock;
      spriteOrBlock = {};
    }

    this.resultManager.startGroup(name, spriteOrBlock.sprite);
    block!();
    this.resultManager.closeGroup(
      spriteOrBlock.summary,
      spriteOrBlock?.visibility as Visibility,
    );
  }
}
