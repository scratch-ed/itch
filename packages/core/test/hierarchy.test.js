import { GroupLevel } from '../src/testplan/hierarchy';
import { GroupedResultManager } from '../src/grouped-output';
import { TabLevel } from '../src/testplan';

test('new flow works', () => {
  const results = [];
  const output = new GroupedResultManager((r) => results.push(r));
  const g = new GroupLevel(output);

  g.group('Groep 1', { sprite: 'Test', visibility: 'show' }, () => {
    g.group('Groep 2', () => {
      g.test('Test 1')
        .feedback({
          correct: 'Hallo',
          wrong: 'No correct at all!',
        })
        .expect('No')
        .toBe('Hallo');
    });
  });

  expect(results).toMatchSnapshot();
});

test('old flow works', () => {
  const results = [];
  const output = new GroupedResultManager((r) => results.push(r));
  const g = new TabLevel(output);

  g.tab('Groep 1', (l) => {
    l.describe('Groep 2', (l) => {
      l.test('Test 1', (l) => {
        l.expect('No')
          .with({
            correct: 'Hallo',
            wrong: 'No correct at all!',
          })
          .toBe('Hallo');
      });
    });
  });

  expect(results).toMatchSnapshot();
});

test('old flow is equivalent to new flow', () => {
  const oldResults = [];
  const output = new GroupedResultManager((r) => oldResults.push(r));
  const t = new TabLevel(output);

  t.tab('Groep 1', (l) => {
    l.describe('Groep 2', (l) => {
      l.test('Test 1', (l) => {
        l.expect('No')
          .with({
            correct: 'Hallo',
            wrong: 'No correct at all!',
          })
          .toBe('Hallo');
      });
    });
  });

  const newResults = [];
  const newOutput = new GroupedResultManager((r) => newResults.push(r));
  const g = new GroupLevel(newOutput);

  g.group('Groep 1', () => {
    g.group('Groep 2', () => {
      g.test('Test 1')
        .feedback({
          correct: 'Hallo',
          wrong: 'No correct at all!',
        })
        .diff()
        .expect('No')
        .toBe('Hallo');
    });
  });

  expect(oldResults).toStrictEqual(newResults);
});
