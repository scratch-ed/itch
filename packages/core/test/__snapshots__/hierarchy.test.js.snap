// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new flow works 1`] = `
[
  {
    "command": "start-judgement",
    "version": 2,
  },
  {
    "command": "start-group",
    "name": "Groep 1",
    "sprite": "Test",
    "tags": [],
    "visibility": "show",
  },
  {
    "command": "start-group",
    "name": "Groep 2",
    "sprite": undefined,
    "tags": [],
    "visibility": "show",
  },
  {
    "command": "start-test",
    "name": "Test 1",
    "tags": [],
  },
  {
    "command": "close-test",
    "feedback": "No correct at all!",
    "status": "wrong",
  },
  {
    "command": "close-group",
    "summary": undefined,
  },
  {
    "command": "close-group",
    "summary": undefined,
  },
]
`;

exports[`old flow works 1`] = `
[
  {
    "command": "start-judgement",
    "version": 2,
  },
  {
    "command": "start-group",
    "name": "Groep 1",
    "sprite": undefined,
    "tags": [],
    "visibility": "show",
  },
  {
    "command": "start-group",
    "name": "Groep 2",
    "sprite": undefined,
    "tags": [],
    "visibility": "show",
  },
  {
    "command": "start-test",
    "name": "Test 1",
    "tags": [],
  },
  {
    "actual": "No",
    "command": "append-diff",
    "expected": "Hallo",
  },
  {
    "command": "close-test",
    "feedback": "No correct at all!",
    "status": "wrong",
  },
  {
    "command": "close-group",
    "summary": undefined,
  },
  {
    "command": "close-group",
    "summary": undefined,
  },
]
`;
